### DISCUSSION TITLE
Implementing a custom predicate?

### DISCUSSION THREAD

**Discussion by ecclarke42** [diesel-rs/diesel#2905] [Q&A] [answered]
In a case where a certain table has more than one identifying column (e.g. a user with id and unique email), I wanted to try to implement a custom predicate to filter for either one of the two.

For example, I'd love to be able to do something like this:

```rust
use crate::schema::users;

// Queryable, etc.
pub struct User {
    pub id: i32, // In practice, this is a id newtype
    pub email: String, 
    ...
}

pub enum UserIdentifier<'a> {
    Id(i32),
    Email(&'a str),
}

fn read(conn: &db::Connection, id: UserIdentifier) -> QueryResult<Self> {
    users::table.filter(id).first::<User>(conn)
}

```

rather than having to differentiate between `users::id.eq(id)` and `users::email.eq(email)`, where some `FilterDsl` implementation handles the translation of `id` to each of those more diesel-y predicates.

But I'm not sure what the types on that impl would look like. There doesn't appear to be a `Predicate` trait to implement. And while it might be possible to `impl FilterDsl<UserIdentifier> for users::table`, that wouldn't cover any join query sources...

If feels possible, but maybe this is just one of those type shenanigans that isn't?

**Comment by weiznich** [comment, marked as answer]
I would suggest to either have an method on `UserIdentifier` which converts that type into a [`Box<dyn BoxableExpression<…, SqlType = Bool>>`](https://docs.diesel.rs/1.4.x/diesel/expression/trait.BoxableExpression.html) or to directly implement [`AsExpression<Bool>`](https://docs.diesel.rs/1.4.x/diesel/expression/trait.AsExpression.html) for `UserIdentifier` (Note: You need to manually implement this. The derive will not work for this case). Both allow reusing all existing `FilterDsl` implementations + allow to mix this with other predicates.

**Reply by ecclarke42** [reply]
`AsExpression` sounds like what I was looking for! But I'm guessing I'll still need to box something in the `match` on `UserIdentifier`, since the `Eq` expressions generated by each of those inner predicates have a different concrete type?

Edit: based on looking that the `expressions` module a bit closer, I should have noticed [this](https://github.com/diesel-rs/diesel/blob/09f8bcd78e3fb634ed69c66f46695b821a5b3822/diesel/src/expression/mod.rs#L312) before asking.
